### PR TITLE
Bug 2060837: resolves multiple channel heads in merged catalogs

### PR DESCRIPTION
**Pull request by jpower432**
# Description

This PR fixes a bug in catalog merging which results in multiple channel heads. It also adds validation during catalog processing to verify the validity of FBCs during the mirroring process.

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [X] Merge unit tests modified to include scenarios with overwritten channel heads
- [X] Manually tested the 4.9 catalogs with different tags

# Testing Plan

1. Spin up two registries 
2. Copy the older catalog to the connected registry
`skopeo copy docker://registry.redhat.io/redhat/redhat-operator-index:v4.9-1646041330 docker://localhost:5000/fake/redhat-operator-index:v4.9`
4. Run oc-mirror with below config
```
apiVersion: mirror.openshift.io/v1alpha2
kind: ImageSetConfiguration
storageConfig:
  local:
    path: metadata
mirror:
    operators:
    - catalog: localhost:5000/fake/redhat-operator-index:v4.9
      headsOnly: false # References latest version of each operator in catalog (true is the default value and can be omitted)
      packages:
        - name: sriov-network-operator
          channels:
            - name: '4.9'
        - name: cluster-logging
          channels:
            - name: 'stable'
```
`./bin/oc-mirror --config ~/configs/imageset-config.yaml docker://localhost:5001 --dest-use-http --log-level=debug --source-use-http`
6. Copy the newer catalog to the same image name to the connected registry
`skopeo copy docker://registry.redhat.io/redhat/redhat-operator-index:v4.9 docker://localhost:5000/fake/redhat-operator-index:v4.9`
7. Run oc-mirror again with the same config

Note: Must test with both commits
- Outcome with the first commit
```
Error: could not build index model from declarative config: invalid index:
├── invalid package "sriov-network-operator":
│   ├── invalid channel "4.9":
│   │   └── multiple channel heads found in graph: sriov-network-operator.4.9.0-202202150216, sriov-network-operator.4.9.0-202202211206
│   └── invalid channel "stable":
│       └── multiple channel heads found in graph: sriov-network-operator.4.9.0-202202150216, sriov-network-operator.4.9.0-202202211206
└── invalid package "cluster-logging":
    ├── invalid channel "stable":
    │   └── multiple channel heads found in graph: cluster-logging.5.3.4-13, cluster-logging.5.3.5-20
    ├── invalid channel "stable-5.2":
    │   └── multiple channel heads found in graph: cluster-logging.5.2.7-18, cluster-logging.5.2.8-21
    └── invalid channel "stable-5.3":
        └── multiple channel heads found in graph: cluster-logging.5.3.4-13, cluster-logging.5.3.5-20
```
- Outcome with second commit - Successfully catalog rebuild

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by afflom** [comment]
@jpower432 The imageset-config api version is alpha2

**Comment by jpower432** [comment]
@afflom yes, this was built from main first to fix the issue for 4.11 and then it will be cherry-picked to the 4.10 branch where the API is still v1alpha1.

**Comment by jharmison-redhat** [comment]
Performed the following test w/ the [integration environment repo](https://github.com/jharmison-redhat/oc-mirror-e2e) directly:

```
$ cat example/vars/scenario.yml
---
cluster_name: disco
openshift_version: "4.9"

scenario_stubs:
  metadata_backend: registry
  mirror_method: to_registry
  registry_type: docker_registry

oc_mirror_source:
  repo: https://github.com/jpower432/bundle.git
  version: bug/multiple-channel-heads

imageset_config_override: |
  apiVersion: mirror.openshift.io/v1alpha2
  kind: ImageSetConfiguration
  storageConfig:
    local:
      path: metadata
  mirror:
      ocp:
        channels:
          - name: stable-4.9
      operators:
      - catalog: registry.internal.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9
        headsOnly: false # References latest version of each operator in catalog (true is the default value and can be omitted)
        packages:
          - name: sriov-network-operator
            channels:
              - name: '4.9'
          - name: cluster-logging
            channels:
              - name: 'stable'
```

1. `make run ANSIBLE_PLAYBOOKS=create`
2. `skopeo login registry.disco.redhat4govaws.io --tls-verify=false  -u admin -p $(cat example/output/registry_admin_password)`
3. `skopeo copy --remove-signatures --dest-tls-verify=false docker://registry.redhat.io/redhat/redhat-operator-index:v4.9-1646041330 docker://registry.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9`
4. `make run ANSIBLE_PLAYBOOKS=test`
5. `skopeo copy --remove-signatures --dest-tls-verify=false docker://registry.redhat.io/redhat/redhat-operator-index:v4.9 docker://registry.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9`

SSH'd to disconnected bastion and confirmed the old catalog parsed the old packages:

```
[ec2-user@bastion ~]$ oc get packagemanifest
NAME                     CATALOG   AGE
cluster-logging                    35m
sriov-network-operator             35m
[ec2-user@bastion ~]$ oc get catalogsource -n openshift-marketplace -oyaml
apiVersion: v1
items:
- apiVersion: operators.coreos.com/v1alpha1
  kind: CatalogSource
  metadata:
    creationTimestamp: "2022-03-14T20:26:18Z"
    generation: 1
    name: redhat-operator-index
    namespace: openshift-marketplace
    resourceVersion: "31849"
    uid: cbe7e25c-2ce9-4ace-ab27-8a38b439f8c7
  spec:
    image: registry.internal.disco.redhat4govaws.io/oc-mirror/catalog/redhat-operator-index:v4.9
    sourceType: grpc
  status:
    connectionState:
      address: redhat-operator-index.openshift-marketplace.svc:50051
      lastConnect: "2022-03-14T21:02:34Z"
      lastObservedState: READY
    registryService:
      createdAt: "2022-03-14T20:29:53Z"
      port: "50051"
      protocol: grpc
      serviceName: redhat-operator-index
      serviceNamespace: openshift-marketplace
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
$ sudo dnf -y install jq
[... SNIP ...]
$ oc get packagemanifest cluster-logging -ojson | jq -r .status.channels[0].currentCSV
cluster-logging.5.3.4-13
```

SSH'd to connected mirror host and performed the following:
1. `cd mirror-content`
2. `~/oc-mirror/bin/oc-mirror --config ~/imageset-config.yml docker://registry.internal.disco.redhat4govaws.io/oc-mirror`

Returned to bastion:
```
$ oc delete pod redhat-operator-index-p6sh8 -n openshift-marketplace
pod "redhat-operator-index-p6sh8" deleted
$ oc get pod -n openshift-marketplace
NAME                                    READY   STATUS    RESTARTS   AGE
marketplace-operator-5777f8869b-bjqfv   1/1     Running   0          50m
redhat-operator-index-bx4ls             1/1     Running   0          15s
$ oc get packagemanifest
NAME                     CATALOG   AGE
cluster-logging                    50m
sriov-network-operator             50m
$ oc get packagemanifest cluster-logging -ojson | jq -r .status.channels[0].currentCSV
cluster-logging.5.3.5-20
```

tl;dr:

lgtm

**Comment by jpower432** [comment]
@soltysh PTAL (priority 2). Thanks!

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by jpower432** [comment]
@afflom Can you please cancel the hold if you are OK with this PR?
